File: LeetCode/628. Maximum Product of Three Numbers/Solution.java
class Solution {
    public int maximumProduct(int[] nums) {
        
        
        int max=Integer.MIN_VALUE;
        int maxi=0;
        int min=Integer.MAX_VALUE;
        int mini=0;
        
        for(int i=0;i<nums.length;i++)
        {
            if(nums[i]>max)
            {
                max=nums[i];
                maxi=i;
            }
            if(nums[i]<min)
            {
                min=nums[i];
                mini=i;
            }
        }
        int secondmin=Integer.MAX_VALUE;
        for(int i=0;i<nums.length;i++)
        {
            if(nums[i]<=secondmin)
            {
                if(i!=mini)
                {
                    secondmin=nums[i];
                }
                    
            }
        }
        
        int secondmax=Integer.MIN_VALUE;
        int smax=0;
        int thirdmax=secondmax;
        
        for(int i=0;i<nums.length;i++)
        {
            if(nums[i]>secondmax && i!=maxi){
                secondmax=nums[i];
                smax=i;}
        }
        
        for(int i=0;i<nums.length;i++)
        {
            if(nums[i]>thirdmax && i!=maxi && i!=smax)
                thirdmax=nums[i];
        }
        
        if(min*secondmin >= secondmax*thirdmax && max>0)
        {
            return min*secondmin*max;
        }
        else
            return secondmax*thirdmax*max;
        
    }
}

File: LeetCode/628. Maximum Product of Three Numbers/Solution.py
class Solution:
    def maximumProduct(self, nums) -> int:
        max1, max2, max3 = float('-inf'), float('-inf'), float('-inf')
        min1, min2 = float('inf'), float('inf')
        
        for n in nums:
            if n <= min1:
                min2 = min1
                min1 = n
            elif n <= min2:
                min2 = n
            if n >= max1:
                max3 = max2
                max2 = max1
                max1 = n
            elif n >= max2:
                max3 = max2
                max2 = n
            elif n >= max3:
                max3 = n
        return max(max1*max2*max3, min1*min2*max1)